homu/quay1.py

```diff
@@ -0,0 +1,325 @@
+        if val is None:
+            raise AttributeError(name)
+        else:
+            return val
+
+    def as_dict(self):
+        return utils.merge_dicts(self.inherited, self.details)
+
+    def delete(self):
+        if self.thing_url is not None:
+            r = requests.delete(self.thing_url,
+                                headers=self.auth_header)
+            r.raise_for_status()
+        else:
+            raise TypeError('Object does not support deletion')
+
+    def refresh(self):
+        if self.thing_url is not None and hasattr(self, 'update_details'):
+            r = requests.get(self.thing_url, headers=self.auth_header)
+            r.raise_for_status()
+            self.update_details(r.json())
+        else:
+            raise TypeError('Object does not support refreshing')
+
+    def _create_child(self, kls, *args, **kwargs):
+        params = utils.merge_dicts(self.basic_params,
+                                   self.inherited)
+        params = utils.merge_dicts(params, parent_url=self.thing_url)
+        params = utils.merge_dicts(params, kwargs)
+        return kls(*args, **params)
+
+    def _make_url(self, *paths, query=None):
+        path = utils.join_paths(*paths)
+        return utils.make_url(self.scheme, self.address,
+                              path, query=query)
+    def _api_url(self, *paths, query=None):
+        paths = (self.api_prefix, ) + paths
+        return self._make_url(*paths, query=query)
+    def _child_api_url(self, *paths, **query_params):
+        if self.thing_url is not None:
+            return utils.add_url_params(utils.join_url(self.thing_url, *paths),
+                                        **query_params)
+        else:
+            raise TypeError('Thing can not have child entities')
+    def _set_thing_api_url(self, parent_url=None, *paths):
+        if parent_url:
+            self.thing_url = utils.join_url(parent_url, *paths)
+        else:
+            self.thing_url = None
+
+class NamespacedThing(QuayThing):
+    def __init__(self, namespace=None, **kwargs):
+        super().__init__(**kwargs)
+        self.inherited['namespace'] = namespace
+
+class Namespace(NamespacedThing):
+    # TODO: save private?
+    def create_repo(self, name, private=True):
+        r = requests.post(self._api_url('repository'),
+                          headers=self.auth_header,
+                          json={'namespace': self.namespace, 'repository': name,
+                                'visibility': 'private' if private else 'public',
+                                'description': ''})
+        r.raise_for_status()
+        info = r.json()
+        return self.repo(info['name'])
+
+    # Get the repo object (NOTE: no requests are made here)
+    def repo(self, name):
+        return self._create_child(Repo, name)
+
+class Repo(NamespacedThing):
+    def __init__(self, name, **kwargs):
+        super(self.__class__, self).__init__(**kwargs)
```

Comment:
warn: using self.__class__ in super might result in something like RuntimeError: maximum recursion